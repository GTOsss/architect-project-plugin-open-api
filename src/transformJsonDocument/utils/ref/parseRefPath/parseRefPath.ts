import { isString } from '#/common/types/common';

/**
 * Превращает $ref строку в массив.
 * См. "./jsonToObject.md" для информации о возможных форматах ссылки.
 * @example
 * parseRefPath('first/second') // → ['first', 'second']
 * parseRefPath('#/first/second') // → ['first', 'second'] */
export const parseRefPath = (ref: string) => {
  if (isString(ref)) {
    const segments = ref.split('/');
    const firstSegment = segments[0];

    if (firstSegment === '#') {
      // ↓ Ссылка на текущий jsonDocument: "#/some/path" ↓

      return segments.slice(1);
    } else if (['http:', 'https:'].includes(firstSegment)) {
      // ↓ Ссылка http/https: "http://some/path" ↓

      throw new Error(`$ref as HTTP/HTTPS url currently not supported: ${ref}`);
    } else {
      // ↓ Обычная ссылка: "some/path" ↓

      return segments;
    }
  } else {
    throw new Error('Invalid ref format. Ref values must be an string');
  }
};
