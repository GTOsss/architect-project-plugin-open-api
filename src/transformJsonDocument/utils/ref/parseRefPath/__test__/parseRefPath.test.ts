import { parseRefPath } from '../parseRefPath';

describe('parseRefPath', () => {
  it('first/second', () => {
    expect(parseRefPath('first/second')).toEqual(['first', 'second']);
  });

  it('#/first/second', () => {
    expect(parseRefPath('#/first/second')).toEqual(['first', 'second']);
  });

  it('https://first/second', () => {
    expect(() => parseRefPath('https://first/second')).toThrowErrorMatchingSnapshot();
  });
});
