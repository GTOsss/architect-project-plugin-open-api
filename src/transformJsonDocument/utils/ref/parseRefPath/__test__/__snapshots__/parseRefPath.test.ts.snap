// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`parseRefPath https://first/second 1`] = `"$ref as HTTP/HTTPS url currently not supported: https://first/second"`;
